### DISCUSSION TITLE
Using Vercel+Nextjs redirect/mask URL only for facebook

### DISCUSSION THREAD

**Discussion by awalonn** [vercel/community#952] [Help] [answered]
Hi guys, hope you'all doing well.
This might be out of topic here, but im hoping someone might be able to help out on this matter.
First of all im new to these stuff like Vercel, Next.js and similar, but what got me into these things was a post i saw on a facebook page, where the person shares his content of his website using vercel domain, but that's not actualy all about,  the url displayed under the featured image was in this format:'example.vercel.app', when i click on it at first facebook shows (example.vercel.app) url in the header, then redirects to his actual domain name, like example.com/post/id.
I've been searching google, but maybe i dont know how to even search for what this guy has done, if someone could help me, i appreciate it so much. thank you, here's a video for better understanding `https://youtu.be/5QZ5wgfyFtU` 

the fb page in question, just for reference so you might figure out what's it all about
`https://www.facebook.com/animalsworld2019s`

i'm just thrilled as How come it only redirects when one clicks on the Facebook link. But when the url is put directly into the address bar it stays on the Vercel site?

the url im talking about, u can copy source link of some post on this fanpage, and u will get something like this
`https://animalsworldsx.vercel.app/hera/?cnt=37956` sharing this url, (u can also use facebook debug tool) it will read everything as `animalsworldsx.vercel.app`, but once u click (from that fb post), redirects to its `domain.com/post-id`!
- facebook debug tool results
`https://i.postimg.cc/YqFzfYQB/Untitled.jpg`

another weird thing, if u share this url to twitter for example,tumblr or even by just copy-paste to url bar-enter, it wont redirect to actual domain, but will remain vercel, im lost!

I've asked on reddit, and youtube , nobody can figure out what is happening, bcoz this is not a 'redirect' or url masking as u might think at first sight! theres something deeper and idk how to doit.

**Comment by amyegan** [comment, marked as answer]
Great question @awalonn! Difficult to say for sure without knowing if this was behavior they actually intended. They may have accidentally shared the `example.vercel.app` link instead of their custom domain, which would redirect to the custom domain like you're seeing. 

Or they might have something configured strangely with their [custom domain](https://vercel.com/docs/concepts/projects/domains). Or maybe they're using [`rewrites`](https://nextjs.org/docs/api-reference/next.config.js/rewrites) and [`redirects`](https://nextjs.org/docs/api-reference/next.config.js/redirects)  (also available in [Vercel config](https://vercel.com/docs/project-configuration#project-configuration/rewrites) for non-Next.js apps) or [Middleware](https://nextjs.org/docs/advanced-features/middleware).

If you can reach out to the owner of the site you're referencing, I think they will be able to give you the most accurate explanation 😄

**Reply by awalonn** [reply]
thanks for ur reply @amyegan 
actually, their behavior is not accidentally , because they post daily on their fanpage posts from their website, using `example.vercel.app` 
i dont want to repeat all what i said above in my question/request, if you only visited that fanpage which i noted above, you can see it yourself what is it all about.

also i dont know to do this, that's why i asked if someone could help me out, the owner of that fanpage doesn't want to share anything at all. Nor does responds to msgs.

- Obviously it uses rewrites and redirects, but how, he is using Next.js u can clearly see that from visiting the base vercel domain they made 'animalsworldsx.vercel.app', it will display the Next.js template.
Anyways, if ur willing to help me out, please check it out again the first message above on this thread, i've included every possible detail for u to understand and review. Thank you

**Reply by amyegan** [reply]
I see the example but do not have access to the source code or know the intentions of the original developer. Without seeing the source code, I'm not sure exactly how `animalsworldsx.vercel.app` is configured to have this behavior. 

One possibility is [Edge Middleware](https://vercel.com/docs/concepts/functions/edge-middleware), which allows for more complex logic in rewrites/redirects. Another possibility is a [matching pattern](https://nextjs.org/docs/api-reference/next.config.js/redirects#header-cookie-and-query-matching) on the config to redirect when the address matches. If you are trying to replicate this behavior, those are the things I would recommend using.

You would need to reach out to the owner of the site to learn exactly what they are doing to cause the behavior. They should have access to code and be able to tell you what they have configured.

**Reply by awalonn** [reply]
nah forget that owner already told you, he's not willing to tell.
Same, as here, nobody willing to try and 'replicate' at least what is done. If i knew howto doit i would not be here asking for help. (im new to vercel, nextjs stuff). i wanted todo this for my website, its wordpress. i wanetd todo this for my post links, but Thanks anyway.

**Reply by amyegan** [reply]
I would recommend using the linked documentation to implement a redirect. The docs contain example code that should help you build an appropriate solution for your site.
- [Create a Next.js App](https://nextjs.org/learn/basics/create-nextjs-app)
- [Redirects](https://nextjs.org/docs/api-reference/next.config.js/redirects) and [Query Matching](https://nextjs.org/docs/api-reference/next.config.js/redirects#header-cookie-and-query-matching)
- [Middleware](https://nextjs.org/docs/advanced-features/middleware)

If you are not using Vercel or Next.js for your site, then you will need to find a solution specific to Wordpress. Sorry, I don't know what else to share!

**Reply by awalonn** [reply]
alright, look what i found
adding this code inside index.tsx at (pages/index.tsx) it redirects `example.vercel.app` to `urldestination.com`

`	
**(this code at very top)**
import type { GetServerSidePropsContext, NextPage } from "next";

**(and this code at the bottom)**
export default Home;

export async function getServerSideProps(context: GetServerSidePropsContext) {
  console.log(context.req.headers);

  const { referer } = context.req.headers;
  if (referer?.includes("facebook.com")) {
    return {
      redirect: {
        destination: "urldestination.com",
        permanent: false,
      },
    };
  }

  // Pass data to the page via props
  return { props: {} };
`}``
 as u can see from the code if referrer is facebook, the url will be displayed and remain as vercel, but once u click it redirects to my site url as defined within the code.
 but, this is working only for the main vercel domain, and i want to create custom posts within this project in order  tor edirect each one to their appropriate post-url on my website, i hope u udnerstand the intention at least, lol.

- like:
'example.vercel.app/post-id redirect to
'mysite.com/post-id'

thank u Amy

**Reply by vireakp** [reply]
did you get solution brother? I'm very appreciate if you can Share us.

**Reply by awalonn** [reply]
> did you get solution brother? I'm very appreciate if you can Share us.

yes and i've abandond this method  already, there was no increase on facebook reach,  tried for a month and nothing happened.

**Comment by learntoearn321** [comment]
i am trying replicate exactly that. can anyone provide any help?

**Reply by vishalkrveer** [reply]
How to make square image ? any guide

**Reply by mohitsahunitrr** [reply]
Use ad manager method for that

**Reply by vishalkrveer** [reply]
Is there any tutorial ? or can you write the steps?
